### ISSUE TITLE
Need to unify approach for associating image filename with detector names

### ISSUE THREAD

**Issue by joelvbernier** [HEXRD/hexrdgui#414]
I am still having the occasional issue with image loading.  Currently, I have a single-detector instrument with the detector name `ge5`; I get an error when I try to load `CeO2_0pt3s_000024.ge5` with the Load Data panel, but I can load it with the File --> Open --> Images dialog.

Is there a reason why we can't simply use the `in` (or `find`, or `count`) built-in to detect the occurrence of the detector name in the input image filenames?  There seems to be no need to mess with different regex for our use case, all we need to know is if the detector name string is anywhere in the image filename; stem, prefix, suffix, whatever.  Hell, we could even have a selector for ignoring case sensitivity.

At any rate, the imports should both behave the _same_ (there is still the need to handle the special case where the images get stored to different directory names that match the detector names, which I believe the "APS Imageseries" button does?) and need to be more robust.

High priority fix.

**Comment by bnmajor** [comment]
@joelvbernier I believe the reason regex was used was to handle a case where a directory contained more than one file with that detector name so that the correct file would be selected. For example, if `first_ge1.h4` was selected for the `ge1` detector `first_ge2.h5` would be selected for the `ge2` detector, but `second_ge2.h5` would be ignored (if it was located in the same directory as `first_ge2.h5`. If this would never be the case then you are correct, this can be greatly simplified with `in` (or an equivalent) and it will just return the first file in the directory that contains that detector name.

Either way the load panel should still find matching files using extension, so that is definitely a bug that needs to be fixed.

**Comment by psavery** [comment]
@bnmajor I think the quick loader uses a regex that is different from the load panel. We might be able to make them both use the load panel way of finding images.

**Comment by joelvbernier** [comment]
Yes, using the `in` approach only works if the user has specified the files they want to load one way or another.  The default behavior can be to just force the user to explicitly select all the image filenames.  While that is fine for simple instruments with a small number of detectors, I can see cases (_e.g._, a 24 detector case we have) where it would definitely be preferred to specify a `glob` pattern instead, _e.g._`sample_data_000012*`, to yield the list of matching filenames.  FWIW, the image_series spec in the HEDM config currently supports globbing.

There are exceptions, however.  In the case where the different detector images have different name stems (it happens, and neither loader can currently handle these to my knowledge), we'll just have to default to the user explicitly selecting all the images in the loader dialog.  Worse still is the case where the detector key(s) are _not_ in the image filenames, which requires explicit selections _and_ association with the correct detector.

I propose that we unify the filename selection process for the direct loader and HEDM imageseries loader, and have it provide the user 3 options:  
- load by explicit selection with auto-association (_e.g._, using `in` comparison on the filename list);
- load using glob pattern (standard wildcards) with auto-association; and
- load by explicit selection with manual association (fallback)
I like the table yielded by the "File --> Open --> Images" dialog that echos back the image names to be loaded for each detector.  It just needs to be modified to allow the detector/image filename associations to be editable to cover the last use case.

For the "Load Data" dialog in the HEDM setting, the `glob`-based approach can effectively handle the case where the different images are in different subdirectories, removing the need for the checkboxes that are currently employed:
```
In [11]: ls                                                                                               
Au_exptest_000005.ge2          aps_py2.py                     ge1/
Hydra_Aug19.yml                calibrate.py                   ge2/
Hydra_Feb19.yml                calibrate_multipanel.py        ge2_Aug19.yml
PUP_AFRL_aug19_FF.par          convert_instr_config.py        lparm_munge.py
Ruby_line_ff_000017_ge1.npz    convert_instr_config.py~       materials.hexrd
Ruby_line_ff_000017_ge2.npz    fastpar_PUP_AFRL_aug19_FF.par  new-instrument-0.yml
Ruby_line_ff_000017_ge3.npz    findorientations.py*           run_archiving.py
Ruby_line_ff_000017_ge4.npz    fitgrains.py*                  run_archiving.py~
Ruby_line_ff_hydra.yml         fitgrains_loop.py*             run_ff_alignment.py
aps.py                         fitgrains_loop.py~*

In [12]: glob.glob('Ruby_line_ff_000017_*')                                                               
Out[12]: 
['Ruby_line_ff_000017_ge1.npz',
 'Ruby_line_ff_000017_ge2.npz',
 'Ruby_line_ff_000017_ge3.npz',
 'Ruby_line_ff_000017_ge4.npz']

In [13]: ['ge4' in i for i in glob.glob('Ruby_line_ff_000017_*')]                                         
Out[13]: [False, False, False, True]

In [14]: glob.glob('*/test_00001*')                                                                       
Out[14]: ['ge1/test_00001.ge1', 'ge2/test_00001.ge2']

In [15]: ['ge1' in i for i in _]                                                                          
Out[15]: [True, False]
```
Sorry to be a pain, but we encounter all these different use cases in reality. Thoughts?

**Comment by bnmajor** [comment]
@joelvbernier Absolutely, we want to handle the all of the use cases and minimize confusing requirements. I want to make sure that I am understanding the cases that we need to handle and how much is actually changing - 

My understanding is that whether the user uses quick load (the `File` -> `Open` -> `Images` path) or the load panel (for the `HEDM` workflow), the files should go through (up to) three checks:
1. Explicit selection: The selected files are checked to be sure that each of the detector names are contained in one of the files
a. Currently the quick load option enforces that only one file be selected per detector, but the load panel allows multiple files per detector. For both cases each detector should have the same number of files. This change should be a matter of allowing explicit file selection through the load panel the way that the quick load does (this is currently not allowed) and simplifying the check to use `in` rather than `regex` comparisons.
2. Load using `glob` pattern: If the explicit selection check has failed (the user did not specify files for each detector), the files that were selected should then be used to search for matches.
a. For example, if detectors `ge1 ge2 ge3 g4` are loaded and the file `Ruby_line_ff_000017_ge1.npz` is selected, we would use the pattern `Ruby_line_ff_000017_*.npz` to find the files for each detector. This is essentially the way that the load panel currently operates, but is not functionality that the quick load has.
b. The search pattern could be manually specified if that is preferred. If this is to be added we would need to consider how the user indicates that they would like to enter a search as well as provide an input space to do so. This would require additional options in the GUI.
c. If no user input is needed, the only changes this would require would be using `glob` to search for files rather than `regex` in order to simplify the search function and improve code readability.
3. Explicit selection with manual association: If the explicit selection check has failed, and the `glob` pattern search has also failed, the user will be provided the option to manually associate their selected files with each detector.
a. This should re-use the current confirmation dialog that appears when images are loaded through the `File` menu. The only change to the dialog would be to allow users to manually select which file is associated with which detector (I am thinking with a simple combo box selector).
b. The only difference between the load panel and the quick load would be that the load panel currently allows for more than one file per detector, whereas the quick load enforces one file per detector. At this point I assume that that will remain the same.
c. Using these three checks the load should only fail if there are not the same number of files for each detector by the end (this would be one per detector for quick load, one or more per detector for the load panel).

> For the "Load Data" dialog in the `HEDM` setting, the `glob`-based approach can effectively handle the case where the different images are in different sub-directories, removing the need for the check boxes that are currently employed:
> 
- The checkbox can certainly be removed with this approach, that would make things much less frustrating when you go through the load process only to realize you've forgotten to tick/un-tick a box...

If any of this seems incorrect or missing something, please let me know and we can get that straightened out. Thanks!

**Comment by joelvbernier** [comment]
@bnmajor yep, looks good!

**Comment by joelvbernier** [comment]
Seems to be an issue here:  I was trying to load a PXRDIP data example to get set up for you and the load selector seems broken.  Two things:
1. It doesn't seem to auto-detect the associated detector names from the filenames anymore; and
2. when I manually select the detector associates, they are not set correctly.
Observe:
![load_from_image_files_broken](https://user-images.githubusercontent.com/1154130/90456574-281ad580-e0ae-11ea-80dc-dce3e88af48f.gif)
I do notice that if I just leave the file table in the defualt order when I open the loader, then associate the detectors manually, it works.
![Screen Shot 2020-08-17 at 5 24 24 PM](https://user-images.githubusercontent.com/1154130/90456738-919ae400-e0ae-11ea-98e1-99fe646756b7.png)

**Comment by bnmajor** [comment]
@joelvbernier This should be fixed in #456
![quick-load-association](https://user-images.githubusercontent.com/51238406/90535697-31a54b80-e149-11ea-938e-50c3182988b3.gif)

**Comment by joelvbernier** [comment]
@bnmajor, weren't we going to have it auto-associate detectors, then fall back to manual associate if that failed?

**Comment by bnmajor** [comment]
@joelvbernier Yes, that should be the behavior and seems to be the case for other instruments. Looking at your examples I found that the issue was using `in` to find a single character detector name - it was finding the detector name in other parts of the path and not making the correct association. #470 should fix this. I tested these changes as thoroughly as I could and I've uploaded a zipped file to the shared directory (`examples/file_testing/test.zip`) that contains all of the sample directories and sub-directories (and all the naming conventions) that I tested through both quick load and the load panel.
